### ISSUE TITLE
🙏Help Wanted - Using Keycloak to auth in Backstage

### ISSUE THREAD

**Issue by BJulliot** [backstage/backstage#27351]
### 📜 Description

TLDR : I'm trying to setup Backstage to setup Keycloak authentification.

So I'm actually trying to setup Backstage to setup a Keycloak OpenId auth.
I followed a mix of two documentation : 
- https://levelup.gitconnected.com/spotifys-backstage-oidc-authentication-using-keycloak-2a664c5009d9
- https://backstage.io/docs/auth/oidc/

This is my app-config.yaml configuration for it : 
```
auth:
  environment: development
  session:
    secret: ${SESSION_SECRET}
  providers:
    Keycloak:
      development:
        clientId: ${AUTH_MY_CLIENT_ID}
        clientSecret: ${AUTH_MY_CLIENT_SECRET}
        metadataUrl: ${MY_KEYCLOAK_URL}
```
In the frontend, I did this : 
```
const providers: SignInProviderConfig[] = [
  {
    id: 'keycloak',
    title: 'Keycloak',
    message: 'Sign in using Keycloak',
    apiRef: oidcAuthApiRef,
  },
];
const app = createApp({
  plugins: [
    grafanaPlugin,
    argocdPlugin,
  ],
  components: {
    SignInPage: props => (
      <SignInPage
        {...props}
        auto
        providers={providers}
      />
    ),
  },
  apis,
  bindRoutes({ bind }) {
    bind(catalogPlugin.externalRoutes, {
      createComponent: scaffolderPlugin.routes.root,
      viewTechDoc: techdocsPlugin.routes.docRoot,
      createFromTemplate: scaffolderPlugin.routes.selectedTemplate,
    });
    bind(apiDocsPlugin.externalRoutes, {
      registerApi: catalogImportPlugin.routes.importPage,
    });
    bind(scaffolderPlugin.externalRoutes, {
      registerComponent: catalogImportPlugin.routes.importPage,
      viewTechDoc: techdocsPlugin.routes.docRoot,
    });
    bind(orgPlugin.externalRoutes, {
      catalogIndex: catalogPlugin.routes.catalogIndex,
    });
  },
  
});
```
I have this apis.ts that is imported into the app.tsx : 
```
import {
  ScmIntegrationsApi,
  scmIntegrationsApiRef,
  ScmAuth,
} from '@backstage/integration-react';
import {
  AnyApiFactory,
  configApiRef,
  createApiFactory,
  ApiRef,
  createApiRef,
  OpenIdConnectApi,
  ProfileInfoApi,
  BackstageIdentityApi,
  SessionApi,
  discoveryApiRef,
  oauthRequestApiRef,
} from '@backstage/core-plugin-api';
import { OAuth2 } from '@backstage/core-app-api';

export const oidcAuthApiRef: ApiRef<
  OpenIdConnectApi & 
  ProfileInfoApi & 
  BackstageIdentityApi & 
  SessionApi 
> = createApiRef({
  id: 'keycloak',
});

export const apis: AnyApiFactory[] = [
  createApiFactory({
    api: oidcAuthApiRef,
    deps: {
      discoveryApi: discoveryApiRef,
      oauthRequestApi: oauthRequestApiRef,
      configApi: configApiRef,
    },
    factory: ({ discoveryApi, oauthRequestApi, configApi }) => {
      console.log('Initializing OIDC API with:', {
        discoveryApi,
        oauthRequestApi,
        configApi,
      });
      return OAuth2.create({
        discoveryApi,
        configApi,
        oauthRequestApi,
        provider: {
          id: 'keycloak',
          title: 'Keycloak',
          icon: () => null,
        },
        environment: configApi.getOptionalString('auth.environment'),
        defaultScopes: ['openid','profile','email'],
        popupOptions: {
          size: {
            fullscreen: true,
          },
        },
      });
    },
  }),
  createApiFactory({
    api: scmIntegrationsApiRef,
    deps: { configApi: configApiRef },
    factory: ({ configApi }) => {
      return ScmIntegrationsApi.fromConfig(configApi);
    },
  }),
  ScmAuth.createDefaultApiFactory(),
];
```

And this is the backend part, into index.ts : 
```
/*
 * Hi!
 *
 * Note that this is an EXAMPLE Backstage backend. Please check the README.
 *
 * Happy hacking!
 */

import { createBackend } from '@backstage/backend-defaults';
import { createBackendModule } from '@backstage/backend-plugin-api';
import {
  authProvidersExtensionPoint,
  createOAuthProviderFactory,
} from '@backstage/plugin-auth-node';
import { oidcAuthenticator } from '@backstage/plugin-auth-backend-module-oidc-provider';
import { DEFAULT_NAMESPACE, stringifyEntityRef } from '@backstage/catalog-model';

const myAuthProviderModule = createBackendModule({
  pluginId: 'auth',
  moduleId: 'keycloak',
  register(reg) {
    reg.registerInit({
      deps: { providers: authProvidersExtensionPoint },
      async init({ providers }) {
        console.log('Registering Keycloak provider'); 

        providers.registerProvider({
          providerId: 'keycloak',
          factory: createOAuthProviderFactory({
            authenticator: oidcAuthenticator,
            async signInResolver(info, ctx) {
              const userRef = stringifyEntityRef({
                kind: 'User',
                name: info.result.fullProfile.userinfo.name as string,
                namespace: DEFAULT_NAMESPACE,
              });
              return ctx.issueToken({
                claims: {
                  sub: userRef,
                  ent: [userRef], 
                },
              });
            },
          }),
        });
      },
    });
  },
});
const backend = createBackend();
backend.add(import('@backstage/plugin-app-backend/alpha'));
backend.add(import('@backstage/plugin-proxy-backend/alpha'));
backend.add(import('@backstage/plugin-scaffolder-backend/alpha'));
backend.add(import('@backstage/plugin-techdocs-backend/alpha'));
backend.add(import('@backstage/plugin-scaffolder-backend-module-github'));
// auth plugin

backend.add(import('@backstage/plugin-auth-backend'));


// See https://backstage.io/docs/backend-system/building-backends/migrating#the-auth-plugin
//backend.add(import('@backstage/plugin-auth-backend-module-guest-provider'));
// See https://backstage.io/docs/auth/guest/provider

// catalog plugin
backend.add(import('@backstage/plugin-catalog-backend/alpha'));
backend.add(
  import('@backstage/plugin-catalog-backend-module-scaffolder-entity-model'),
);
backend.add(import('@backstage/plugin-kubernetes-backend'));
// See https://backstage.io/docs/features/software-catalog/configuration#subscribing-to-catalog-errors
backend.add(import('@backstage/plugin-catalog-backend-module-logs'));

// permission plugin
backend.add(import('@backstage/plugin-permission-backend/alpha'));
// See https://backstage.io/docs/permissions/getting-started for how to create your own permission policy
backend.add(
  import('@backstage/plugin-permission-backend-module-allow-all-policy'),
);

// search plugin
backend.add(import('@backstage/plugin-search-backend/alpha'));

// search engine
// See https://backstage.io/docs/features/search/search-engines
backend.add(import('@backstage/plugin-search-backend-module-pg/alpha'));

// search collators
backend.add(import('@backstage/plugin-search-backend-module-catalog/alpha'));
backend.add(import('@backstage/plugin-search-backend-module-techdocs/alpha'));

backend.add(myAuthProviderModule);

backend.start();

```
With this configuration, and the one I have in my keycloak, when I try to log by clicking on SIGN IN, there is a popup that instantly dissapear and this error message appear : 
```
Login failed; caused by Error: Authentication rejected, iss missing from the response
```

### 👍 Expected behavior

The expected behavior : 
It should display on the popup my keycloak login page

### 👎 Actual Behavior with Screenshots

A popup appear and dissapear instantly, leaving this message into the backstage page : 
Login failed; caused by Error: Authentication rejected, iss missing from the response

![image](https://github.com/user-attachments/assets/f79e0f6b-b644-40fc-9185-b510e0a215a2)

### 👟 Reproduction steps

This is from a fresh backstage installation : 
-Setup a new keycloak client.
-Follow a mix of those two documentation : 
  -  https://levelup.gitconnected.com/spotifys-backstage-oidc-authentication-using-keycloak-2a664c5009d9
  - https://backstage.io/docs/auth/oidc/
  
 Which result in the file I send

### 📃 Provide the context for the Bug.

Can not log using Keycloak OpenId Auth

### 🖥️ Your Environment

I'm using firefox version : 128.3.1esr
The yarn backstage-cli info returned : 
```
yarn backstage-cli info
OS:   Linux 6.9.9-arch1-1 - linux/x64
node: v20.18.0
yarn: 4.4.1
cli:  0.27.1 (installed)
backstage:  1.31.3

Dependencies:
  @backstage/app-defaults                                          1.5.11
  @backstage/backend-app-api                                       1.0.0
  @backstage/backend-common                                        0.25.0
  @backstage/backend-defaults                                      0.5.0
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.1.18
  @backstage/backend-plugin-api                                    1.0.0, 1.0.1
  @backstage/catalog-client                                        1.7.0, 1.7.1
  @backstage/catalog-model                                         1.5.0, 1.7.0, 1.6.0-next.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.8
  @backstage/cli                                                   0.27.1
  @backstage/config-loader                                         1.9.1
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.15.0, 1.15.1
  @backstage/core-compat-api                                       0.2.8, 0.3.0, 0.3.1
  @backstage/core-components                                       0.12.5, 0.14.10, 0.15.0, 0.15.1, 0.15.1-next.0
  @backstage/core-plugin-api                                       1.10.0, 1.10.0-next.1, 1.9.4
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.9
  @backstage/frontend-app-api                                      0.10.0
  @backstage/frontend-defaults                                     0.1.1
  @backstage/frontend-plugin-api                                   0.7.0, 0.8.0, 0.9.0
  @backstage/frontend-test-utils                                   0.2.1
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.32, 1.2.0
  @backstage/integration                                           1.15.0, 1.15.1
  @backstage/plugin-api-docs                                       0.11.10
  @backstage/plugin-app-backend                                    0.3.75
  @backstage/plugin-app-node                                       0.1.25
  @backstage/plugin-app                                            0.1.1
  @backstage/plugin-auth-backend-module-atlassian-provider         0.3.0
  @backstage/plugin-auth-backend-module-auth0-provider             0.1.0
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.2.0
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.2.0
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.2.0
  @backstage/plugin-auth-backend-module-bitbucket-server-provider  0.1.0
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.3.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.3.0
  @backstage/plugin-auth-backend-module-github-provider            0.2.0
  @backstage/plugin-auth-backend-module-gitlab-provider            0.2.0
  @backstage/plugin-auth-backend-module-google-provider            0.2.0
  @backstage/plugin-auth-backend-module-guest-provider             0.2.0
  @backstage/plugin-auth-backend-module-microsoft-provider         0.2.0
  @backstage/plugin-auth-backend-module-oauth2-provider            0.3.0
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.2.0
  @backstage/plugin-auth-backend-module-oidc-provider              0.3.0
  @backstage/plugin-auth-backend-module-okta-provider              0.1.0
  @backstage/plugin-auth-backend-module-onelogin-provider          0.2.0
  @backstage/plugin-auth-backend                                   0.23.0
  @backstage/plugin-auth-node                                      0.5.2, 0.5.3
  @backstage/plugin-auth-react                                     0.1.6
  @backstage/plugin-bitbucket-cloud-common                         0.2.23
  @backstage/plugin-catalog-backend-module-logs                    0.1.1
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.2.0
  @backstage/plugin-catalog-backend                                1.26.1
  @backstage/plugin-catalog-common                                 1.1.0
  @backstage/plugin-catalog-graph                                  0.4.10
  @backstage/plugin-catalog-import                                 0.12.4
  @backstage/plugin-catalog-node                                   1.13.0, 1.13.1
  @backstage/plugin-catalog-react                                  1.14.0, 1.13.1
  @backstage/plugin-catalog                                        1.23.1
  @backstage/plugin-events-node                                    0.4.0
  @backstage/plugin-home-react                                     0.1.17
  @backstage/plugin-home                                           0.7.11
  @backstage/plugin-kubernetes-backend                             0.18.7
  @backstage/plugin-kubernetes-common                              0.8.0, 0.8.3
  @backstage/plugin-kubernetes-node                                0.1.20
  @backstage/plugin-kubernetes-react                               0.4.0, 0.4.4
  @backstage/plugin-kubernetes                                     0.11.16
  @backstage/plugin-org                                            0.6.30
  @backstage/plugin-permission-backend-module-allow-all-policy     0.2.0
  @backstage/plugin-permission-backend                             0.5.49
  @backstage/plugin-permission-common                              0.7.14, 0.8.1
  @backstage/plugin-permission-node                                0.8.3, 0.8.4
  @backstage/plugin-permission-react                               0.4.26, 0.4.27
  @backstage/plugin-proxy-backend                                  0.5.6
  @backstage/plugin-scaffolder-backend-module-azure                0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.2.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.3.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.2.0
  @backstage/plugin-scaffolder-backend-module-gitea                0.2.0
  @backstage/plugin-scaffolder-backend-module-github               0.5.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.5.0
  @backstage/plugin-scaffolder-backend                             1.25.0
  @backstage/plugin-scaffolder-common                              1.5.6
  @backstage/plugin-scaffolder-node                                0.4.11
  @backstage/plugin-scaffolder-react                               1.12.1
  @backstage/plugin-scaffolder                                     1.25.1
  @backstage/plugin-search-backend-module-catalog                  0.2.2
  @backstage/plugin-search-backend-module-pg                       0.5.35
  @backstage/plugin-search-backend-module-techdocs                 0.2.2
  @backstage/plugin-search-backend-node                            1.3.2
  @backstage/plugin-search-backend                                 1.5.17
  @backstage/plugin-search-common                                  1.2.14
  @backstage/plugin-search-react                                   1.8.0
  @backstage/plugin-search                                         1.4.17
  @backstage/plugin-signals-react                                  0.0.5
  @backstage/plugin-techdocs-backend                               1.10.13
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.15
  @backstage/plugin-techdocs-node                                  1.12.11
  @backstage/plugin-techdocs-react                                 1.2.8
  @backstage/plugin-techdocs                                       1.10.10
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.13
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.6.0, 1.7.0
  @backstage/theme                                                 0.2.19, 0.5.7, 0.6.0
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.10-next.0, 1.0.10, 1.0.9
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by camilaibs** [comment]
Hi @BJulliot, have you tried asking on the [Backstage Discord](https://discord.gg/backstage-687207715902193673) server? It looks like there are other community members using Keycloak there, maybe they can share with you how they set up their Backstage.

**Comment by carlosthe19916** [comment]
@BJulliot if you still didn't find a working configuration of Backstage+Keycloak I've just managed to do it here https://github.com/carlosthe19916/backstage

**Comment by afaras72** [comment]
@carlosthe19916 I have used your code for my implementation but still get **"Login failed; caused by Error: Authentication rejected, login_required"** any idea why one could be getting this error?

**Comment by MatheusEble** [comment]
@afaras72 try add prompt

```
auth:
  session:
    secret: XXXXXXX
  environment: development
  providers:
    guest: {}
    keycloak:
      development:
        metadataUrl: XXXXXXX
        clientId: XXXXXXX
        clientSecret: XXXXXXX
        prompt: auto
```

**Comment by Nithursika** [comment]
@afaras72 
I also tried, but I'm getting the same error:

"Login failed; caused by Error: Authentication rejected, login_required"

Did you ever figure out what was causing it in your case?
